aten/src/ATen/native/cuda/ScatterGatherKernel.cu

```diff
@@ -65,17 +65,27 @@ __global__ void _scatter_gather_elementwise_kernel(int N, func_t f) {
   }
 }
 
-template <int nt, int vt, typename func_t>
+template <typename func_t>
 static void _launch_scatter_gather_kernel(int64_t N, const func_t& f) {
   TORCH_INTERNAL_ASSERT(N >= 0 && N <= std::numeric_limits<int32_t>::max());
   if (N == 0) {
     return;
   }
 
+  const int nt = at::cuda::warp_size() * 2;
+  constexpr  int vt = 4;
   const dim3 block(nt);
-  const dim3 grid((N + block.x * vt - 1) / (block.x * vt));
+  const dim3 grid((N + nt * vt - 1) / (nt * vt));
   const auto stream = at::cuda::getCurrentCUDAStream();
-  _scatter_gather_elementwise_kernel<nt, vt, func_t><<<grid, block, 0, stream>>>(N, f);
+
+  if (nt == 64) {
+    _scatter_gather_elementwise_kernel<64, vt, func_t>
+        <<<grid, block, 0, stream>>>(N, f);
+  } else if (nt == 128) {
+    _scatter_gather_elementwise_kernel<128, vt, func_t>
+        <<<grid, block, 0, stream>>>(N, f);
+  }
+
```

Comment:
Should we guard this with ifdefs on ROCm? This increases the CUDA binary size.

Reply:
Call/launch sites are precisely the problem this PR is trying to address.  Launch sites must become dynamic and query the warp size at runtime.  `at::cuda::warp_size()`.  `C10_WARP_SIZE` is 32 for CUDA, and up to this point has been 64 for HIP.  For HIP, we must now support 32 and 64 depending on the gfx target.  The `C10_WARP_SIZE` for HIP was recently changed from 64 to `warpSize` that comes from HIP headers.  This does correct the warpSize used during compilation of kernels, but does not address the issue of querying the hardware at runtime.  We need a `num_threads()` that is a constexpr for compilation, but we need a dynamic `query_num_threads()` or `runtime_num_threads()` that perhaps wraps `at::cuda::warp_size() * 4`.

Reply:
Do you actually want different number of threads depending on the warp size? Most call sites are for elementwise kernels that don't care about relationship to warp size, we specify num_threads() as 4*C10_WARP_SIZE, but what we really want is 128. Should we just modify `num_threads()` to not depend on the warp size?

Reply:
@ngimel If it is acceptable to have num_threads not rely on the warps size. Then that works for us.

Reply:
We have ifdef these changes so that they only affect rocm. It has a minor impact on our stack and we will address it in a subsequent pr. We will create an issue to track it.